iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/DirectoryCopySourceCheck.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.checks;
+
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.iac.common.api.checks.CheckContext;
+import org.sonar.iac.common.api.checks.IacCheck;
+import org.sonar.iac.common.api.checks.InitContext;
+import org.sonar.iac.docker.tree.api.AddTree;
+import org.sonar.iac.docker.tree.api.CopyTree;
+import org.sonar.iac.docker.tree.api.ParamTree;
+import org.sonar.iac.docker.tree.api.SyntaxToken;
+import org.sonar.iac.docker.utils.SyntaxTokenUtils;
+
+@Rule(key = "S6470")
+public class DirectoryCopySourceCheck implements IacCheck {
+
+  private static final String MESSAGE_CURRENT_OR_ROOT = "Make sure that recursively copying directories is safe here.";
+  private static final String MESSAGE_GLOBING = "Make sure that using globbing in a %s source is safe here.";
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(AddTree.class, DirectoryCopySourceCheck::checkAdd);
+    init.register(CopyTree.class, DirectoryCopySourceCheck::checkCopy);
+  }
+
+  private static void checkAdd(CheckContext ctx, AddTree add) {
+    for (SyntaxToken src : add.srcs()) {
+      String path = SyntaxTokenUtils.sanitize(src);
+      if (!path.startsWith("http://") && !path.startsWith("https://")) {
+        reportIfSensitive(ctx, src, isSensitivePath(path), "ADD");
+      }
+    }
+  }
+
+  private static void checkCopy(CheckContext ctx, CopyTree copy) {
+    if (hasOption(copy.options(), "from")) return;
+
+    for (SyntaxToken src : copy.srcs()) {
+      reportIfSensitive(ctx, src, isSensitivePath(SyntaxTokenUtils.sanitize(src)), "COPY");
+    }
+  }
+
+  private static boolean hasOption(List<ParamTree> options, String key) {
+    return options.stream().anyMatch(param -> param.name().equals(key));
+  }
+
+  private static void reportIfSensitive(CheckContext ctx, SyntaxToken src, PathSensitivity sensitivity, String instructionName) {
+    if(sensitivity == PathSensitivity.ROOT_OR_CURRENT) {
+      ctx.reportIssue(src, MESSAGE_CURRENT_OR_ROOT);
+    } else if (sensitivity == PathSensitivity.TOP_LEVEL_GLOBBING) {
+      ctx.reportIssue(src, String.format(MESSAGE_GLOBING, instructionName));
+    }
+  }
+
+  enum PathSensitivity {
+    SAFE, ROOT_OR_CURRENT, TOP_LEVEL_GLOBBING
+  }
+
+  /**
+   * Indicate if a path is sensitive. A path is sensitive if any of the current condition is true :
+   * - if the provided path resolve to a root level or to the current directory
+   * - if there is any top level entry (level <= 1) ending with a wildcard *
+   * Examples of root level : '/' ; '/test/..' ; 'c:/' ; 'c:/test/..'
+   * Examples of current level : '.' ; './test/..'
+   * Examples of top level entry ending with wildcard : 'a*' ; './a*' ; '/a*' ; './test/../a*'
```

Comment:
To make it prettier for JavaDoc we need to use tags:
```suggestion
  /**
   * Indicate if a path is sensitive. A path is sensitive if any of the current condition is true :
   * <ul>
   * <li>if the provided path resolve to a root level or to the current directory</li>
   * <li>if there is any top level entry (level <= 1) ending with a wildcard</li>
   * </ul>
   * Examples of root level : '/' ; '/test/..' ; 'c:/' ; 'c:/test/..'
   * Examples of current level : '.' ; './test/..'
   * Examples of top level entry ending with wildcard : 'a*' ; './a*' ; '/a*' ; './test/../a*'
   */
```